syntax = "proto3";

import "app/models/proto/shared.proto";
import "buf/validate/validate.proto";

enum NoteStatus {
  open = 0;
  closed = 1;
  hidden = 2;
}

service NoteService {
  rpc GetNote(GetNoteRequest) returns (GetNoteResponse) {
    option idempotency_level = NO_SIDE_EFFECTS;
  }
}

message GetNoteRequest {
  uint64 id = 1 [(buf.validate.field).uint64.gt = 0];
}

message GetNoteResponse {
  NoteData note = 1 [(buf.validate.field).required = true];
}

message NoteData {
  message Header {
    optional User user = 1;
    uint64 created_at = 2;
    string body_rich = 3;
  }

  uint64 id = 1;
  LonLat location = 2 [(buf.validate.field).required = true];
  NoteStatus status = 3;
  Header header = 4 [(buf.validate.field).required = true];
  bool is_subscribed = 5;
  optional uint32 disappear_days = 6;
}

message NoteCommentPage {
  message Comment {
    enum Event {
      opened = 0;
      closed = 1;
      reopened = 2;
      commented = 3;
      hidden = 4;
    }

    optional User user = 1;
    Event event = 2;
    uint64 created_at = 3;
    string body_rich = 4;
  }

  repeated Comment comments = 1;
}

message NoteCommentResult {
  NoteData note = 1 [(buf.validate.field).required = true];
  NoteCommentPage comments = 2 [(buf.validate.field).required = true];
}

message RenderNotesData {
  message Note {
    uint64 id = 1;
    LonLat location = 2 [(buf.validate.field).required = true];
    string text = 3;
    NoteStatus status = 4;
  }

  repeated Note notes = 1;
}
